### PR TITLE
feat: implement /operations/assets/:asset_issuer/:asset_code

### LINKED ISSUE #392: horizon: implement /operations/:asset_code/:asset_issuer
it would be nice to have an endpoint exposed by horizon to retrieve all the operations regarding a particular assetCode/issuer pair.

### PR DESCRIPTION
**Pull request by phra**
fixes #392

### CODE REVIEW

**Comment by phra** [comment]
a test related to effects is failing.. do you think is due to the changes of this PR?

**Comment by bartekn** [comment]
Sounds similar to https://github.com/stellar/go/issues/396

**Comment by phra** [comment]
@bartekn the only thing that pops in my mind is that some scenario are not reset correctly after some tests.. i've changed the scenario for some operations tests as you can see in the diff.. any idea about fixing it?

**Comment by phra** [comment]
CI is green! :heavy_check_mark: 
Mike Fair ask me to extend the new query with `filter only by asset_issuer and not by asset_code` case, i will work on it.

**Comment by phra** [comment]
@bartekn can you review the changes? thanks.

File: services/horizon/internal/db2/history/operation.go [hunk 1]
```diff
@@ -105,6 +105,39 @@ func (q *OperationsQ) ForTransaction(hash string) *OperationsQ {
 	return q
 }
 
+// ForAsset filters the query to a only operations for a specific asset,
+// specified by the the asset code and the asset issuer address.
+func (q *OperationsQ) ForAsset(assetIssuer string, assetCode string) *OperationsQ {
+	q.sql = q.sql.Where(
+		"hop.details @> json_object(ARRAY['asset_code', ? ,'asset_issuer', ? ])::jsonb"+
+			" OR hop.details @> json_object(ARRAY['buying_asset_code', ? ,'buying_asset_issuer', ? ])::jsonb"+
+			" OR hop.details @> json_object(ARRAY['selling_asset_code', ? ,'selling_asset_issuer', ? ])::jsonb",
```
[the remark this answers is not in the document]
Comment (phra) [reply services/horizon/internal/db2/history/operation.go]: for sure! jsonb columns should be indexable btw.

**Comment by nullstyle** [comment]
This endpoint, as currently designed is not something I'm willing to merge.

It's simply unclear to me what "it would be nice to have an endpoint exposed by horizon to retrieve all the operations regarding a particular assetCode/issuer pair" concretely means.  For example:  Why aren't `allow_trust` operations being included in the result set?  Surely they do "regard" the asset being trusted.  This issue is a blocker.  

Other issues:

-  `/operations/assets/:asset_issuer/:asset_code`  violates our conventions for building endpoints in horizon.  
  - An asset at present can only be specified on the query string using the `type/issuer/code` triplet.  Before we start representing assets in the path we need to agree upon the format.  
  - The nesting is wrong even is we did have that agreed upon format;  If anything it should be `/assets/:asset/operations`
- No docs 😛

**Comment by MikeFair** [comment]
I've got an interest in seeing this functionality integrated and have helped others asking how to achieve similar results (though most folks are looking for the all the current trustlines of an asset; some of the people, like myself, also need operations/payments/effects for an issuer's asset(s)).

> Why aren't allow_trust operations being included in the result set?

I definitely agree any operation involving the asset ought to be included (change_trust, allow_trust, mange_offer, payment, path_payment, all of them); `/payments` could be similarly extended to involve only operations creating a transfer of the asset.

> what "it would be nice to have an endpoint exposed by horizon to retrieve all the operations regarding a particular assetCode/issuer pair" concretely means.

As for the utility of getting "operations involving an asset", issuers and workflow authors can find all sorts of great uses for this kind of a filtered feed.  When you're the issuer of an asset and you're writing code for the community of folks dealing in that asset, getting a filtered stream of operations for the asset is exceedingly useful.  The same goes for the `/payments` and `/effects` feeds, at which point it seems weird to exclude `/transactions` but I'm not going to make a fuss over it.

Some examples off the top of my head:
* if you have a rule that says "Only Authorized Accounts can make offers to sell/buy this asset", there's no native way to enforce an "authorized buyers/sellers" list (if you're trying to create a "Service Coin" controlling which accounts can deal in the side of the coin you need helps prevent fraud).  Revoking the trustline authorization of any account that executes a manage_offer operation to buy/sell the asset (breaking the Issuer's rule) is one way to create that effect.  
* If you "require authorization", seeing the "change_trust" operation can inform you of the account id to begin the authorization workflow.  
* If you are trying to create a visualization graph for how the asset has changed hands over its lifetime, the feed of operations filtered just for that asset makes your life much easier.  
* If you are trying to set up some kind of "smart contract" that uses the transfer of an asset from one account to another to cause some kind of reactionary effect to happen somewhere else in the ledger or even off chain (e.g. if an asset is being used to track what account currently controls the ownership or authorized access to another system (like a 'dealer button' in a poker game), seeing the asset transfer operation come through can trigger updates in the other systems; I have a process where an account sending an asset to another account triggers updating a trustline on the issuer account. Horizon makes this kind of detection cumbersome as inspecting ledger events by asset is not an exposed concept).

More generally, "Accounts" that hold many "Assets" are how individual holders might see and view the utility of the ledger; but on the service/asset issuers side the worldview is reversed.  The "Asset" has many "Trustors", or clients. The Issued Asset Object has many "balances" that are its Holders.  It also has "offers".  It'd be great if it could have "custom data" too.  Many folks have noticed a lot of utility in creating a "type 2" account object for creating a registered assets concept in the system.  Issuers and Application authors, care about the many accounts transacting in their asset.  Horizon currently offers little assistance to them on this front.

> The nesting is wrong even is we did have that agreed upon format; If anything it should be /assets/:asset/operations

I definitely prefer this approach too.  It mirrors `/accounts/:accountid/operations` which reinforces the idea of "assets" as a first class object in the system.

> An asset at present can only be specified on the query string using the `type/issuer/code` triplet. 

* `/assets/native/operations`?
* while alphanum4 and alphanum12 are mutually exclusive types, keeping type in the asset id is prudent.  
`/assets/credit_alphanum4/:asset_issuer/:asset_code/operations`
`/assets/credit_alphanum12/:asset_issuer/:asset_code/operations`

Is there any symbol you could see approving to mean a "wildcard" for the asset_code?  This way an Issuer could easily query for all the operations of all the assets it issues processed through the same code base?

This assets path fits perfectly with another feature many folks need/want:
`/assets/type/:asset_issuer/:asset_code` <--- An "account like" result for the asset
`/assets/type/:asset_issuer/:asset_code/balances` <--- All the current holders of an asset
`/assets/type/:asset_issuer/:asset_code/balances/:accountid` <--- A specific holder's trustline

> I have a bad feeling about performance of new queries. 

* A table could be created like the effects/operations_participants table for accounts if this really becomes an issue, 
* custom expression indexes can absolutely reach inside jsonb columns
* Creating an "assets" table like accounts to give Assets a numeric 'id' would help a lot

One thing to keep in mind is that if they are being used, then people are finding them useful and finding a way to make the lookups fast is worth exploring; if they are somewhat slow, but provide functionality that is useful to some, not possible without the queries, and less frequently used, then the performance impact is not often actualized.

This feature, and the feature to look up all the holders of an asset by issuer or issuer/asset_code, are the two biggest missing features I think Horizon should be providing that I inform people they have to run their own Horizon node and add their own queries to get.

**Comment by phra** [comment]
@MikeFair thanks for the exhaustive explanation.

> Why aren't allow_trust operations being included in the result set?

i'm not filtering the results using a particular operation type so all the operations are retrieved at the moment.

let's agree on an url convention so i can update the PR.

File: services/horizon/internal/db2/history/operation.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (MikeFair) [reply services/horizon/internal/db2/history/operation.go]: I believe the result of the jsonb index is that it index the whole object and not the parts; separate indexes for the field subset to be tested using custom expressions work "better".

**Comment by MikeFair** [comment]
@phra @nullstyle @bartekn 

@nullstyle Thanks for bringing up the points to enhance the feature.

After thinking about it some more, the use cases I've seen, and what information/usage the users and devs have at the time of the initial request, and what they would want to be able to access, I believe:

1) the "triplet" should move `type` to the end, making it: `/assets/:asset_issuer/:asset_code/:asset_type`
End users will be told the issuer/asset_code they are looking for, but the type is more of an 'asset property' than an 'asset identifier'.  Long story short, the only useful 'asset_type' at the moment is the keyword 'native' to distinguish it from the others.  The alphanum4/12 distinction is fairly pointless for identification purposes.  By placing it on the end, it can regularly be left off the path without causing any ambiguity.  The question then is what do to about 'native' which I address below.

2) the `/operations` feature should be available at every level.
  `/assets/operations` <--- would exclude inflation, manage_data, set_options (maybe optional here)
  `/assets/:asset_issuer/operations` <-- all operations for assets involving the specific asset_issuer
  `/assets/:asset_issuer/:asset_code/operations` <-- all operations for all types of that issuer/code pair
  `/assets/:asset_issuer/:asset_code/:asset_type/operations` <-- all operations for the specific asset

3) The intent is to eventually add `/transactions`, `/payments`, and `/effects` endpoints as well

4) The question then becomes how to handle "native"
  a) Make "native" a magic string for the issuer field in the path
  b) Use: asset_issuer = networkId / network master account ; asset_code = XLM; asset_type = 'native'
  c) Create `/assets/native` and `/assets/issued/:asset_issuer/:asset_code/:asset_type` for the paths
  d) Exclude 'Native' from the concept of the `/assets` endpoint ensuring all assets have an issuer/code 
  e) Don't use '/' to make "paths" but instead use an encoding convention for a single path element string
    This would be like `/assets/[:asset_issuer],[:asset_code],[:asset_type]/operations`
    and `/assets/native/operations`
    The format is an 'opaque string encoding convention' that must be 'resolved' to identify an asset(s)

I don't know what the intention is for the future of asset types, but I think they should be removed as part of an asset's identification; for native I'd prefer to see the networkId used as the issuer and 'native' used as the asset_code, or something else that keeps with the convention.

If the intention is for the issuer to reuse the same asset_code and have many "types" of it then this is probably better handled by extending the asset/trustline record data, instead of including the asset_type as a prefix in the asset identifier.  The current types 'credit_alphanum4/12' imply a correlated type 'debit_alphanum4/12'.  I think this would be better handled by having a 'credit_limit', 'credit_balance', and 'debit_limit', 'debit_balance' properties of the trustline.  

I get the necessity of an asset's type to identify the code workflows and rules to apply when processing operations/effects; but I steadfastly disagree that the system should encourage/allow an account to hold the same issuer/asset_code pair and multiple asset_types.  The confusion to users would be endless.  If an account cannot hold multiple of the same issuer/asset code pair then :asset_type becomes a 'property' of the asset/trustline and not a typical path 'identifier' that users are regularly concerned with.

For example, I don't usually distinguish between credit card USD, debit card USD, check USD, and cash USD.  They are all various types of USD, and we don't usually consider those distinct types until we first think of the currency we're looking to deal in.  `United States/USD/credit_card`, and `United States/USD/bank_loan` are far more intuitive than `credit_card/United States/USD`, and `bank_loan/United States/USD`.

It appears to me that forcing the inclusion of the :asset_type into the path ahead of issuer/asset either:
  (1) makes developers have to predict information they cannot know in advance;
  (2) are redundant because the asset_code and issuer are distinct;
  (3) if not distinct, and holder's are allowed to have multiple assets using the same issuer/code pair, then dealing with these assets becomes much more confusing and difficult in practice

Thanks for the consideration,
Mike

File: services/horizon/internal/db2/history/operation.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (phra) [reply services/horizon/internal/db2/history/operation.go]: https://blog.codeship.com/unleash-the-power-of-storing-json-in-postgres/

**Comment by phra** [comment]
@nullstyle 

> Why aren't allow_trust operations being included in the result set? Surely they do "regard" the asset being trusted. This issue is a blocker.

with `/operations/assets/:asset_issuer/:asset_code` you will retrieve every `allow_trust` related to a specific asset using the following [query](https://github.com/stellar/go/pull/398/files#diff-b4ca44f9b397d755f06e2cb7e66409cbR110), instead with `/operations/assets/:asset_issuer` you will retrieve every `allow_trust` related to that issuer using this [one](https://github.com/stellar/go/pull/398/files#diff-b4ca44f9b397d755f06e2cb7e66409cbR128), so also `allow_trust` operation related only to the issuer and not specific to an asset.

can you elaborate more and help me out to understand your question?

_PS_: btw i need this feature in order to avoid to tap in my own horizon db instance and just use horizon apis, so let's discuss on how do we want to proceed with it.

/cc @MikeFair

**Comment by MikeFair** [comment]
One thing I think we can agree on is the endpoint should be changed to
/assets/:asset/operations
And remove the /operations/assets/* references.

And to add one more level, `:asset_type`

That means:
/assets/operations <-- all operations involving an issuer
/assets/:asset_issuer/operations
/assets/:asset_issuer/:asset_code/operations
/assets/:asset_issuer/:asset_code/:type/operations


After that, the only discussion left is what, if anything, to do about the
native asset.

**Comment by phra** [comment]
@MikeFair 

> `/assets/operations` <-- all operations involving an issuer

where is the issuer parameter? do you mean `assets`?

**Comment by MikeFair** [comment]
> @MikeFair <https://github.com/MikeFair>
>
> /assets/operations <-- all operations involving an issuer
>
> where is the issuer parameter? do you mean assets?
>

There isn't one for that endpoint; that literally means all operations for
assets.

**Comment by phra** [comment]
@MikeFair ok.

@nullstyle do you agree with the proposed changes so i can proceed to implement them?

**Comment by phra** [comment]
sorry for the bump, but we really need this feature implemented in horizon.
retrieving operations regarding a specific asset (asset_code + asset_issuer pair) is a fundamental feature and there were often discussions about this topic on the stellar slack.

can someone of the team give out some ideas on how do we want to proceed? /cc @bartekn /cc @nullstyle

**Comment by nullstyle** [comment]
@phra While I have a pending, and long response to all the discussion that has happened, I want to give you a heads up about how to proceed:  I think you should fork horizon and do the work you need to do to unblock yourself.

There are still plenty of design discussions that we need to consider and I simply don't have the time to properly vet your work or integrate into horizon as a whole.  It appears that you've got the most pressing needs for this to be included and I don't think we have any competing solutions in progress;  In the meantime I believe you should maintain your own fork.

**Comment by phra** [comment]
i think we will switch to a graphql endpoint exposing core and horizon dbs in readonly mode using postgraphile like in this repository: https://github.com/TinyAnvil/stellar-graphql-docker

atm it's an easier solution compared to managing our own fork of horizon. _also because we lack go devs_ :man_shrugging:

i will reconsider to use an horizon fork only if we will experience problems with postgraphile.

for the moment let's leave the PR open.

**Comment by nullstyle** [comment]
@phra Using the graphql solution looks like a great idea.

Instead of leaving this PR open, let's instead reopen it when we continue the discussion.  It's effortless to do so and we can get tasks with incomplete technical designs off of the lists we scrub each sprint.

**Comment by nullstyle** [comment]
> The existence of those other tools doesn't the warrant closing this PR.
And at the moment I doubt this PR will get reopened later because I've seen no evidence that old closed PRs ever get revisited.  There's nothing that will tell anyone to come back and look at it.

The search engine will bring people back to the issue, and you and phra are free to continue the conversation here as long as you'd like, of course.  He opened the issue and it sounds like he's got a better path in front of him and it doesn't seem like he'll be contributing much to the code.  I certainly don't have the time to work with him on the issue, unfortunately.  

> How many other closed PRs have promises to get reopened later when people "get around to discussing them"?

To what promises are you referring?  This discussion seems to be taking an accusatory turn and I'm not keen to engage.  Let me say my piece and I'll be out:  While we work to solve everyone's problems and incorporate everyone's feedback, sometimes we can't incorporate a PR and will close it.  By clicking the merge button on a pull request I am vouching for the code; Until I'm happy with it I won't give my vote, and I don't promise my vote until I see what I'm voting on.  That sounds a little too aggressive for my tastes but I don't know of a softer way to be strictly accurate.  We're dealing with money in this project, stuff of real world impact, so some careful scrutiny of any addition seems warranted.

Now, in this specific case we've got no one to push the code forward, so it seems fine to close it.  We can always open it up later when there is someone to build the code and press the project forward.

> or outright reject it and let the community know this isn't a feature you'll take from them

I'll pick the third option; I'll let the feature cook for longer and perhaps on the next attempt it will be up to my personal quality bars with minimal intervention.  Listen, I can understand that not getting your particular need met by the official release is frustrating and I hear and am sorry for my part in that frustration.  We're working to bring more engineers into the team to help improve on our ability to make changes to our fork of horizon, but for the meantime we're a bit (that is to say, lot) short on platform team engineering.  I wish I had more time, skill, eloquence, and social energy to properly engage on every issue up for inclusion into horizon.  I don't and I apologize for that deficit.  I'm still not willing to compromise on the API's quality, as I personally define it.

> How about redoing it to make "asset identifiers" a query string filter on the `/operations` path?

This is the least offensive way, in my opinion, at the moment for filtering an operation by assets.  It doesn't fundamentally change the objections I have with this PR.  

@MikeFair here's what I would recommend:  Open a new issue in which the community can discuss your specific use case.  API design isn't supposed to happen in pull requests.  Attract an audience and come up with a design.  Make reference to this pull request to give other readers a historical path to follow should they care.  From that design issue the feature can proceed.

**Pull request closed without merging**